pandas/io/html.py

```diff
@@ -346,40 +361,149 @@ def _build_doc(self):
         """
         raise AbstractMethodError(self)
 
-    def _build_table(self, table):
-        header = self._parse_raw_thead(table)
-        body = self._parse_raw_tbody(table)
-        footer = self._parse_raw_tfoot(table)
+    def _build_table(self, table_html):
+        header, body, footer = self._parse_raw_thead_tbody_tfoot(table_html)
+        # the above "footer" actually produces a footer. The below "footer"
+        # rarely does. The below "footer" is the legacy behavior and so I'm
+        # leaving it for the time being.
+        footer = self._parse_raw_tfoot(table_html)
         return header, body, footer
 
-    def _parse_raw_thead(self, table):
-        thead = self._parse_thead(table)
+    def _parse_raw_thead_tbody_tfoot(self, table_html):
+        """Given a table, return parsed header, body, and foot.
+           Header and body are lists-of-lists. Top level list is a list of
+           rows. Each row is a list of parsed elements.
+
+           Logic: Use <thead>, <tbody>, <tfoot> elements to identify
+                  header, body, and footer, otherwise:
+                  - Put all rows into body
+                  - Move rows from top of body to header only if
+                    all elements inside row are <th>
+                  - Move rows from bottom of body to footer only if
+                    all elements inside row are <th>
+
+        Parameters
+        ----------
+        table_html : node-like
```

Comment:
Let's take advantage of this time to re-specify this as a proper data type if possible.